### PR TITLE
Make Sage work with maxima 5.47

### PR DESCRIPTION
**Pull request by antonio-rojas**
Maxima 5.47 has been released. The main change that affects Sage is https://sourceforge.net/p/maxima/code/ci/f5ca00582c24cfedca53a664335f6c13e1e40cf9, which makes the maxima ecl library crash because the `e-val` variable is unassigned. We fix this by running the `initialize-runtime-globals` function on load.

The remaining issues are a few harmless test failures caused by numerical noise and output format changes (such as additional parentheses near minus signs)

### CODE REVIEW

File: src/doc/de/tutorial/interfaces.rst
Comment (tornaria) [inline src/doc/de/tutorial/interfaces.rst]:
```suggestion
    sage: A.eigenvectors().sage()
```
seems to produce more stable result, what do you think?

File: src/doc/de/tutorial/interfaces.rst
Comment (tornaria) [inline src/doc/de/tutorial/interfaces.rst]:
```suggestion
    sage: maxima("expr_2: -5*sin(x)*(cos(x/2)*cos(y) + sin(x/2)*sin(2*y)+ 3.0)").sage()
```

File: src/doc/de/tutorial/interfaces.rst
Comment (tornaria) [inline src/doc/de/tutorial/interfaces.rst]:
```suggestion
    [[[0, 4], [3, 1]],
     [[[1, 0, 0, -4], [0, 1, 0, -2], [0, 0, 1, -4/3]], [[1, 2, 3, 4]]]]
```

File: src/doc/de/tutorial/interfaces.rst
Comment (tornaria) [inline src/doc/de/tutorial/interfaces.rst]:
```suggestion
    -5*(cos(1/2*x)*cos(y) + sin(1/2*x)*sin(2*y) + 3.0)*sin(x)
```
same here

File: src/doc/de/tutorial/tour_algebra.rst
Comment (tornaria) [inline src/doc/de/tutorial/tour_algebra.rst]:
```suggestion
    sage: lde1 = de1.laplace("t","s"); lde1.sage()
```
Here as well, the output seems easier to read.

File: src/doc/de/tutorial/tour_algebra.rst
Comment (tornaria) [inline src/doc/de/tutorial/tour_algebra.rst]:
```suggestion
    2*s^2*laplace(x(t), t, s) - 2*s*x(0) + 6*laplace(x(t), t, s) - 2*laplace(y(t), t, s) - 2*D[0](x)(0)
```

File: src/sage/calculus/calculus.py
Comment (tornaria) [inline src/sage/calculus/calculus.py:1339]: What happened here? Answer was `und` and now is `ind`, is there a semantic difference? Who is choosing the naming?

File: src/sage/functions/bessel.py
Comment (tornaria) [inline src/sage/functions/bessel.py]: Much nicer IMO:
```
sage: m = maxima(bessel_J(2, x))
sage: m.integrate(x).sage()
1/24*x^3*hypergeometric((3/2,), (5/2, 3), -1/4*x^2)
```

**Comment by tornaria** [comment]
Just FTR, doctests  pass for me with 5.47.0 (normal and long) with the current PR. My suggestions are merely cosmetic, and I wouldn't mind keeping it as it is  (all these testing is horrible anyway).

Once we settle on something, I could try to run tests on 5.47.0 / 5.46.0 / 5.45.1, just to make sure it's ok.

**Comment by kiwifb** [comment]
If the appending of `.sage()` fix most of the brackets, appearing and disappearing then it is a much better solution. It makes the answers much clearer. 

Another solution would be subtracting the know answer and simplify it to 0. That can only work for exact results of course.

File: src/sage/calculus/calculus.py
Comment (antonio-rojas) [reply src/sage/calculus/calculus.py:1339]: See https://sourceforge.net/p/maxima/code/ci/df1748a88b2906aac866297a2cec48747e2d3728

File: src/doc/en/constructions/linear_algebra.rst
Comment (tornaria) [inline src/doc/en/constructions/linear_algebra.rst]: missed this  one

File: src/sage/functions/orthogonal_polys.py
Comment (tornaria) [inline src/sage/functions/orthogonal_polys.py:977]:
```
sage: maxima(chebyshev_U(2,x, hold=True)).sage()
4*(x - 1)^2 + 8*x - 5
```
I don't know what this is trying to acomplish... I guess that maxima knows how to handle  `chebyshev_U(2, x)`.

File: src/sage/functions/special.py
Comment (tornaria) [inline src/sage/functions/special.py]: Really, it seems to me the previous answer is wrong, and the new answer is correct, even if it's not the nicest (note that `abs(sin(z)) + 2*round(z/pi)` is also incorrect). Let's just write something like
```
sage: elliptic_e(z, 1).simplify()                # not tested: bad answer with maxima < 5.47
2*round(z/pi) - sin(pi*round(z/pi) - z)
```
(is there a way to flag a test as "expected to fail" rather than "not test" ?)

This is an even better test which shows the new answer is ok:
```
sage: f(z) = elliptic_e(z, 1).simplify()
sage: all([abs(f(z).simplify()-elliptic_e(z,1))<1e-15
....:      for z in [RR.random_element(10) for _ in range(10)]])
True
```
you can see this passes with maxima 5.47 but fails with maxima 5.46.

Comment (tornaria) [reply src/sage/functions/special.py]: Could do something like:
```
sage: maxima.version() < '5.47' or all([
....:     abs(f(z).simplify()-elliptic_e(z,1))<1e-15
....:     for z in [RR.random_element(10) for _ in range(100)]])
True
```
but it's a bit ugly :shrug:   (ideal would be a way to label the test something  like `xfail when maxima.version() < '5.47'`.

File: src/sage/interfaces/interface.py
Comment (tornaria) [inline src/sage/interfaces/interface.py]: I give up... :fearful:  At some point these tests are just not useful anymore... with all this `...` stuffed in.

**Comment by tornaria** [comment]
Summary: don't waste too much time in this... Maybe just put the "new correct" answer in the `elliptic_e` test (with the `# not tested b/c old maxima is broken` tag) just so that the thing that shows up in the documentation is correct (and the thing in the comment seems to be incorrect (???).

File: src/doc/en/tutorial/interfaces.rst
Comment (orlitzky) [inline src/doc/en/tutorial/interfaces.rst:324]: These don't actually need to be echoed for the example. I'd suggest e.g.

```
sage: _ = maxima("expr_2: -5*sin(x)*(cos(x/2)*cos(y) + sin(x/2)*sin(2*y)+ 3.0)")
```

instead.

File: src/sage/functions/orthogonal_polys.py
Comment (antonio-rojas) [reply src/sage/functions/orthogonal_polys.py:977]: I guess it's testing conversion to maxima, so I'm not sure converting back to sage the result makes much sense

File: src/doc/en/constructions/linear_algebra.rst
Comment (orlitzky) [reply src/doc/en/constructions/linear_algebra.rst]: I cannot imagine why we would tell the poor user to use the low-level maxima interface to find eigenvectors in a linear algebra tutorial. My opinion: delete the bad advice (and the failing test).

File: src/sage/functions/bessel.py
Comment (orlitzky) [inline src/sage/functions/bessel.py]: I don't think we really care about the exact string output here; we want to know that going to maxima, taking a derivative, and going back to sage gives you the same result that taking the derivative in sage would:

```
sage: x,y = SR.var('x,y')
sage: f = Bessel(typ='K')(x,y)
sage: expected = f.derivative(y)
sage: actual = maxima(f).derivative('_SAGE_VAR_y').sage()
sage: bool(actual == expected)
True
```

File: src/sage/functions/orthogonal_polys.py
Comment (orlitzky) [reply src/sage/functions/orthogonal_polys.py:977]: Yeah, it's supposed to be an example of the `__init__` method. Why send it to maxima at all?

File: src/sage/functions/other.py
Comment (orlitzky) [inline src/sage/functions/other.py]: This looks like it works without the fuzz (because 1/4 is a power of two):

```
sage: a = floor(5.25 + x); a
floor(x + 5.25000000000000)
sage: a.simplify()
floor(x + 0.25) + 5
```

but the best example is probably,

```
sage: a = floor(27/5 + x); a
floor(x + 27/5)
sage: a.simplify()
floor(x + 2/5) + 5
```

File: src/sage/functions/special.py
Comment (orlitzky) [inline src/sage/functions/special.py:458]: I guess this is OK for now. This is testing for a specific maxima bug (3967) that was fixed in 5.47, and that is already part of the maxima test suite. So normally, I would say "delete it," but there aren't many other examples for `elliptic_e` and showing that it can be simplified to a more-explicit form is nice.

File: src/sage/interfaces/interface.py
Comment (orlitzky) [reply src/sage/interfaces/interface.py]: It's still not a great solution, but if we drop the negative sign from `'-cos(x)`, then the answer is consistent.

File: src/sage/interfaces/maxima.py
Comment (orlitzky) [inline src/sage/interfaces/maxima.py:60]: This is more resilient as e.g.

```
sage: x,y = SR.var('x,y')
sage: actual = maxima.factor('x^5 - y^5').sage()
sage: expected = -(y-x)*(y^4+x*y^3+x^2*y^2+x^3*y+x^4)
sage: bool(actual == expected)
```

File: src/sage/interfaces/maxima.py
Comment (orlitzky) [inline src/sage/interfaces/maxima.py:82]: There's no reason to continue using the (complicated) expression `F` here. Something predictable like `maxime(1/2)` should work fine.

File: src/sage/interfaces/maxima_abstract.py
Comment (orlitzky) [inline src/sage/interfaces/maxima_abstract.py]: You can call `.sage()` on this one and compare it to the expected answer with `==`.

File: src/sage/modules/free_module_element.pyx
Comment (orlitzky) [inline src/sage/modules/free_module_element.pyx]: abs tol here?

File: src/sage/symbolic/relation.py
Comment (orlitzky) [inline src/sage/symbolic/relation.py]: abs tol here?

File: src/doc/en/tutorial/interfaces.rst
Comment (orlitzky) [inline src/doc/en/tutorial/interfaces.rst:271]: This is difficult to test correctly because the eigenvalues, eigenvectors, and multiplicities are split across three lists and need to be matched up. The eigenvectors I would not expect to be unique anyway.

Lazy fix: switch it to a matrix with only nonnegative eigenvalues and a set of eigenvectors whose entries can be chosen integral and nonnegative.

File: src/sage/functions/bessel.py
Comment (orlitzky) [inline src/sage/functions/bessel.py:297]: I think this is a pretty useless test. What does anyone care about the maxima interface here?

In any case, `f.integrate()` _already_ uses maxima by default. We could specify `f.integrate(x,algorithm='maxima')` to ensure that the maxima interface is still tested in case the default ever changes... but again, who wants to read about the maxima interface here?

File: src/doc/en/constructions/linear_algebra.rst
Comment (orlitzky) [inline src/doc/en/constructions/linear_algebra.rst:281]: I thought I commented on this one already, but don't see it. I don't know why we're telling people to use the low-level maxima interface to compute the eigenvalues of a 2x2 matrix in a linear algebra tutorial. Personally I would just delete it (and the surrounding explanation).

File: src/doc/en/tutorial/tour_algebra.rst
Comment (orlitzky) [inline src/doc/en/tutorial/tour_algebra.rst:224]: The next paragraph tries to explain the output from this example, and so no longer applies.

But instead of changing it, this is another place where I don't understand why we're telling people to use the maxima interface. Sage supports laplace transforms itself. The (explicit) round trip through maxima and back can be avoided:

```
sage: t,s = SR.var('t,s')
sage: x = function('x')
sage: y = function('y')
sage: f = 2*x(t).diff(t,2) + 6*x(t) - 2*y(t)
sage: f.laplace(t,s)
2*s^2*laplace(x(t), t, s) - 2*s*x(0) + 6*laplace(x(t), t, s) - 2*laplace(y(t), t, s) - 2*D[0](x)(0)
```

File: src/doc/en/constructions/linear_algebra.rst
Comment (antonio-rojas) [reply src/doc/en/constructions/linear_algebra.rst:281]: Then the gap section immediately after that should be dropped too. I think we should leave any further doc cleanup for a follow up ticket.

**Review by orlitzky — APPROVED** [review]
Ok, we shouldn't hold this up any longer. The first commit is pretty important. Thanks.

File: src/sage/interfaces/maxima_lib.py
Comment (tornaria) [inline src/sage/interfaces/maxima_lib.py]: It seems this is unfixing #26968, as seen here: https://github.com/void-linux/void-packages/actions/runs/5366826557/jobs/9736459766?pr=44624#step:7:43488

Presumably, calling `initialize-runtime-globals` will run `set-pathnames` and fail without the workaround that is in the `try` block... Can this be moved *after* that?

Comment (tornaria) [reply src/sage/interfaces/maxima_lib.py]: My suggestion:

```diff
--- a/src/sage/interfaces/maxima_lib.py
+++ b/src/sage/interfaces/maxima_lib.py
@@ -133,12 +133,11 @@ if MAXIMA_FAS:
     ecl_eval("(require 'maxima \"{}\")".format(MAXIMA_FAS))
 else:
     ecl_eval("(require 'maxima)")
-ecl_eval("(maxima::initialize-runtime-globals)")
 ecl_eval("(in-package :maxima)")
-ecl_eval("(setq $nolabels t))")
-ecl_eval("(defvar *MAXIMA-LANG-SUBDIR* NIL)")
 ecl_eval("(set-locale-subdir)")
 
+# this has to happen early so the workaround works
+# do not add other initialization before this block
 try:
     ecl_eval("(set-pathnames)")
 except RuntimeError:
@@ -155,6 +154,8 @@ except RuntimeError:
     # Call `(set-pathnames)` again to complete its job.
     ecl_eval("(set-pathnames)")
 
+ecl_eval("(initialize-runtime-globals)")
+ecl_eval("(setq $nolabels t))")
 ecl_eval("(defun add-lineinfo (x) x)")
 ecl_eval('(defun principal nil (cond ($noprincipal (diverg)) ((not pcprntd) (merror "Divergent Integral"))))')
 ecl_eval("(remprop 'mfactorial 'grind)")  # don't use ! for factorials (#11539)
```

**Comment by tornaria** [comment]
Please merge https://github.com/antonio-rojas/sage/pull/2 into this PR before merging this PR in main.

**Pull request merged**
